File: demo/referral-api.py
import requests
import json

class ReferralAPI:
    def __init__(self, api_key, base_url='https://api.your-domain.com'):
        self.api_key = api_key
        self.base_url = base_url
        self.session = requests.Session()
        self.session.headers.update({
            'X-API-Key': api_key,
            'Content-Type': 'application/json'
        })
    
    def create_referral(self, campaign_id, referrer_id, referee_id=None):
        """Create a new referral"""
        try:
            response = self.session.post(
                f'{self.base_url}/v1/referrals',
                json={
                    'campaignId': campaign_id,
                    'referrerId': referrer_id,
                    'refereeId': referee_id
                }
            )
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as error:
            print(f'Error creating referral: {error}')
            raise error
    
    def track_click(self, referral_code, metadata=None):
        """Track a referral click"""
        try:
            response = self.session.post(
                f'{self.base_url}/v1/referrals/{referral_code}/click',
                json={'metadata': metadata or {}}
            )
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as error:
            print(f'Error tracking click: {error}')
            raise error
    
    def record_conversion(self, referral_code, amount=None, metadata=None):
        """Record a conversion"""
        try:
            response = self.session.post(
                f'{self.base_url}/v1/referrals/{referral_code}/convert',
                json={
                    'amount': amount,
                    'metadata': metadata or {}
                }
            )
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as error:
            print(f'Error recording conversion: {error}')
            raise error
    
    def get_stats(self, campaign_id=None):
        """Get referral statistics"""
        try:
            params = {'campaignId': campaign_id} if campaign_id else {}
            response = self.session.get(f'{self.base_url}/v1/stats', params=params)
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as error:
            print(f'Error fetching stats: {error}')
            raise error

# Usage Example
def main():
    api = ReferralAPI('YOUR_API_KEY')
    
    try:
        # Create a referral
        referral = api.create_referral('campaign_123', 'user_456')
        print('Referral created:', referral)
        
        # Track a click
        click = api.track_click(referral['referralCode'], {
            'userAgent': 'Mozilla/5.0...',
            'ipAddress': '192.168.1.1'
        })
        print('Click tracked:', click)
        
        # Record a conversion
        conversion = api.record_conversion(referral['referralCode'], 99.99, {
            'orderId': 'order_789',
            'productId': 'prod_abc'
        })
        print('Conversion recorded:', conversion)
        
        # Get stats
        stats = api.get_stats()
        print('Stats:', stats)
        
    except Exception as error:
        print('API Error:', error)

if __name__ == '__main__':
    main()

File: demo/referral-api.js
// eslint-disable-next-line @typescript-eslint/no-require-imports
const axios = require('axios');

class ReferralAPI {
    constructor(apiKey, baseURL = 'https://api.your-domain.com') {
        this.apiKey = apiKey;
        this.baseURL = baseURL;
        this.client = axios.create({
            baseURL,
            headers: {
                'X-API-Key': apiKey,
                'Content-Type': 'application/json'
            }
        });
    }

    // Create a new referral
    async createReferral(campaignId, referrerId, refereeId = null) {
        try {
            const response = await this.client.post('/v1/referrals', {
                campaignId,
                referrerId,
                refereeId
            });
            return response.data;
        } catch (error) {
            console.error('Error creating referral:', error.response?.data || error.message);
            throw error;
        }
    }

    // Track a referral click
    async trackClick(referralCode, metadata = {}) {
        try {
            const response = await this.client.post(`/v1/referrals/${referralCode}/click`, {
                metadata
            });
            return response.data;
        } catch (error) {
            console.error('Error tracking click:', error.response?.data || error.message);
            throw error;
        }
    }

    // Record a conversion
    async recordConversion(referralCode, amount = null, metadata = {}) {
        try {
            const response = await this.client.post(`/v1/referrals/${referralCode}/convert`, {
                amount,
                metadata
            });
            return response.data;
        } catch (error) {
            console.error('Error recording conversion:', error.response?.data || error.message);
            throw error;
        }
    }

    // Get referral statistics
    async getStats(campaignId = null) {
        try {
            const params = campaignId ? { campaignId } : {};
            const response = await this.client.get('/v1/stats', { params });
            return response.data;
        } catch (error) {
            console.error('Error fetching stats:', error.response?.data || error.message);
            throw error;
        }
    }
}

// Usage Example
async function main() {
    const api = new ReferralAPI('YOUR_API_KEY');
    
    try {
        // Create a referral
        const referral = await api.createReferral('campaign_123', 'user_456');
        console.log('Referral created:', referral);
        
        // Track a click (when someone clicks the referral link)
        const click = await api.trackClick(referral.referralCode, {
            userAgent: navigator.userAgent,
            ipAddress: '192.168.1.1'
        });
        console.log('Click tracked:', click);
        
        // Record a conversion (when the referee makes a purchase)
        const conversion = await api.recordConversion(referral.referralCode, 99.99, {
            orderId: 'order_789',
            productId: 'prod_abc'
        });
        console.log('Conversion recorded:', conversion);
        
        // Get stats
        const stats = await api.getStats();
        console.log('Stats:', stats);
        
    } catch (error) {
        console.error('API Error:', error);
    }
}

module.exports = ReferralAPI;